### PR TITLE
docs(examples): render custom HTML to PDF

### PR DESCRIPTION
**Pull request by guumaster**
add example to create an HTML page in memory, render it with puppeteer and save it as a pdf file.

### CODE REVIEW

**Comment by guumaster** [comment]
`I signed it!`

File: examples/custom-html-pdf.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright 2017 Google Inc. All rights reserved.
```
Comment (JoelEinbinder) [inline examples/custom-html-pdf.js]: 2018

File: examples/custom-html-pdf.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+const puppeteer = require('puppeteer');
+
+const YOUR_HTML = `<!doctype html>
+<html>
+<head>
+  <meta charset="utf-8">
+  <title>Example HTML</title>
+  <!-- you can link any local/external file -->
+  <link rel="stylesheet" href="css/styles.css">
```
Comment (JoelEinbinder) [inline examples/custom-html-pdf.js:32]: You can't link to an external file if using a data url.

File: examples/custom-html-pdf.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+</body>
+</html>
+`;
+
+(async() => {
+  const browser = await puppeteer.launch();
+  const page = await browser.newPage();
+
+  // set your content and wait until render is finished
+  await page.goto(`data:text/html,${YOUR_HTML}`, { waitUntil: 'networkidle2' });
```
Comment (JoelEinbinder) [inline examples/custom-html-pdf.js]: why wait for networkidle2?

**Review by JoelEinbinder — COMMENTED** [review]
Can you elaborate on a scenario where this would be useful for someone?

File: examples/custom-html-pdf.js (hunk 2 above)
Comment (guumaster) [reply examples/custom-html-pdf.js:32]: I have a working example that loads JS and CSS and Fonts from the network based on this code.

File: examples/custom-html-pdf.js (hunk 1 above)
Comment (guumaster) [reply examples/custom-html-pdf.js]: updated.

File: examples/custom-html-pdf.js (hunk 3 above)
Comment (guumaster) [reply examples/custom-html-pdf.js]: updated to `networkidle0`. I've read the docs wrong.

**Comment by guumaster** [comment]
> Can you elaborate on a scenario where this would be useful for someone?

If you have a server that renders HTML and you want to generate pdf from any page, you can use this code to do it instead of navigating to an internal page with `goto()`.

I've tried with `setContent()/content()` but even if you use `await page.content()` it wont show immediately. So after a while of going through the issues and docs I got this example working.

I thought it would be useful for somebody else.

Thanks for reviewing it!

**Comment by aslushnikov** [comment]
@guumaster The "goto hack" is a consequence of #728; I'd rather have it fixed than spreading the hack further.

#728 requires contributing upstream to Chromium and DevTools protocol; let me know if you want to give it a shot.

**Comment by guumaster** [comment]
Probably I can't contribute to #728 to Chromium.

This example may not be too important. But is not such a hack either, it is something that you can normally do with a browser, and `puppeteer` is meant to control a browser, isn't it?  Maybe someone that needs this can find the code in this issue useful.

Thanks for your time anyway.

PS: You think you can merge it if I get the example working with `setContent()/content()` ?

**Pull request closed without merging**